### ISSUE TITLE
Add support for the forthcoming JS SDK update

### ISSUE THREAD

**Issue by coder11** [fluencelabs/aqua#222]
We are planning a big update of JS SDK. Aqua compiler will need to support following features:

1) Adding `callParams` as the last argument to callback functions
2) Generating type information for the tetraplets structure
3) Generating type information for service definitions
4) Generating functions which should allow create a handler for a service

Please refer to the following code snippet for the example of generated typescript code:

https://github.com/fluencelabs/fluence-js/blob/c196cc16f0c501bec2e40e7b3bbf35d314397d32/src/__test__/unit/compiler.spec.ts

**Comment by coder11** [comment]
Service definition example:

```
service HelloWorld:
    sayHello(s: string)
    getNumber() -> i32
```

What should be generated in TS:

```
function registerHelloWorld(
    client: FluenceClient,
    id: string,
    service: {
        sayHello: (s: string, callParams: CallParams<'s'>) => void;
        getNumber: (callParams: CallParams<null>) => number;
    },
) {
    client.callServiceHandler.use((req, resp, next) => {
        if (req.serviceId !== id) {
            next();
            return;
        }

        if (req.fnName === 'sayHello') {
            const callParams = {
                ...req.particleContext,
                tetraplets: {
                    s: req.tetraplets[0],
                },
            };
            const res = service.sayHello(req.args[0], callParams);
            resp.retCode = ResultCodes.success;
            resp.result = null;
        }

        if (req.fnName === 'getNumber') {
            const callParams = {
                ...req.particleContext,
                tetraplets: {},
            };
            const res = service.getNumber(callParams);
            resp.retCode = ResultCodes.success;
            resp.result = res;
        }

        next();
    });
}
```

**Comment by coder11** [comment]
Function definition example:

```
func callMeBack(callback: string, i32 -> ()):
    callback("hello, world", 42)
```

What should be generated in TS (take a closer look at `callback` paramter:

```
async function callMeBack(
    client: FluenceClient,
    callback: (
        arg0: string, // force format
        arg1: number,
        callParams: CallParams<'arg0' | 'arg1'>,
    ) => void,
    config?: { ttl?: number },
): Promise<void> {
    let request: RequestFlow;
    const promise = new Promise<void>((resolve, reject) => {
        const r = new RequestFlowBuilder()
            .disableInjections()
            .withRawScript(
                `
(xor
 (seq
  (call %init_peer_id% ("getDataSrv" "-relay-") [] -relay-)
  (xor
   (call %init_peer_id% ("callbackSrv" "callback") ["hello, world" 42])
   (call %init_peer_id% ("errorHandlingSrv" "error") [%last_error% 1])
  )
 )
 (call %init_peer_id% ("errorHandlingSrv" "error") [%last_error% 2])
)
            `,
            )
            .configHandler((h) => {
                h.on('getDataSrv', '-relay-', () => {
                    return client.relayPeerId!;
                });

                h.use((req, resp, next) => {
                    if (req.serviceId === 'callbackSrv' && req.fnName === 'callback') {
                        const callParams = {
                            ...req.particleContext,
                            tetraplets: {
                                arg0: req.tetraplets[0],
                                arg1: req.tetraplets[1],
                            },
                        };
                        const res = callback(req.args[0], req.args[1], callParams);
                        resp.retCode = ResultCodes.success;
                        resp.result = {};
                    }
                    next();
                });

                h.onEvent('errorHandlingSrv', 'error', (args) => {
                    // assuming error is the single argument
                    const [err] = args;
                    reject(err);
                });
            })
            .handleScriptError(reject)
            .handleTimeout(() => {
                reject('Request timed out for callMeBack');
            });
        if (config?.ttl) {
            r.withTTL(config.ttl);
        }
        request = r.build();
    });
    await client.initiateFlow(request!);
    return Promise.race([promise, Promise.resolve()]);
}
```

**Comment by alari** [comment]
Fixes #210

**Comment by alari** [comment]
Fixes #197

**Comment by alari** [comment]
Fixed in #251
